trace/trace_test.go

```diff
@@ -437,3 +446,303 @@ func TestSpanContextFromContext(t *testing.T) {
+				assert.Equal(t, tc.expectedTraceState, result)
+			}
+		})
+	}
+}
+
+func TestTraceStateInsert(t *testing.T) {
+	testCases := []struct {
+		name               string
+		traceState         TraceState
+		keyValue           label.KeyValue
+		expectedTraceState TraceState
+		expectedErr        error
+	}{
+		{
+			name: "OK case - add new",
+			traceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+			keyValue: label.String("key4@vendor", "val4"),
+			expectedTraceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key4@vendor", "val4"),
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+		},
+		{
+			name: "OK case - replace",
+			traceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+					label.String("key2", "val2"),
+					label.String("key3", "val3"),
+				},
+			},
+			keyValue: label.String("key2", "valX"),
+			expectedTraceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key2", "valX"),
+					label.String("key1", "val1"),
+					label.String("key3", "val3"),
+				},
+			},
+		},
+		{
+			name: "Invalid key/value",
+			traceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+				},
+			},
+			keyValue: label.String("key!", "val!"),
+			expectedTraceState: TraceState{
+				kvs: []label.KeyValue{
+					label.String("key1", "val1"),
+				},
+			},
+			expectedErr: errInvalidTraceStateKeyValue,
+		},
+		{
+			name:               "Too many entries",
+			traceState:         TraceState{generateKVsWithMaxMembers()},
+			keyValue:           label.String("keyx", "valx"),
+			expectedTraceState: TraceState{generateKVsWithMaxMembers()},
+			expectedErr:        errInvalidTraceStateMembersNumber,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			result, err := tc.traceState.Insert(tc.keyValue)
+			if tc.expectedErr != nil {
+				require.Error(t, err)
+				assert.Equal(t, tc.expectedErr, err)
+				assert.Equal(t, tc.traceState, result)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, tc.expectedTraceState, result)
+			}
+		})
+	}
+}
+
+func TestTraceStateFromKeyValues(t *testing.T) {
+	testCases := []struct {
+		name               string
+		kvs                []label.KeyValue
+		expectedTraceState TraceState
+		expectedErr        error
+	}{
+		{
+			name:               "OK case",
+			kvs:                generateKVsWithMaxMembers(),
+			expectedTraceState: TraceState{generateKVsWithMaxMembers()},
+		},
+		{
+			name:               "OK case (empty)",
+			expectedTraceState: TraceState{},
+		},
+		{
+			name: "Too many entries",
+			kvs: func() []label.KeyValue {
+				kvs := generateKVsWithMaxMembers()
+				kvs = append(kvs, label.String("keyx", "valX"))
+				return kvs
+			}(),
+			expectedTraceState: TraceState{},
```

Comment:
Similar here with regards to truncation. Should this return a TraceState with `keyx=valX` truncated as well as the included error?